File: textanalyzer.java
//Mark Gomes
//CSC 330
//Assignment 1 - Text Parsing
import java.io.File;
import java.util.Scanner;
import java.io.FileNotFoundException;

public class textanalyzer
{
    public static void main(String[] args)
    {
        checkFile(args.length);             //Checks if any filename is given

        String fileName = args[0];
        double wordNum, syllNum, sentNum;
        double alpha, beta, FRI, FKGLI;
        int[] wordData = new int[3];        //0:Words, 1:Syllables, 2:Sentences
        wordData[0] = 0; wordData[1] = 0; wordData[2] = 0;

        countWords(wordData, fileName);

        wordNum = wordData[0];  syllNum = wordData[1];  sentNum = wordData[2];
        alpha = syllNum/wordNum;                            //Calculates the indexes
        beta = wordNum/sentNum;
        FRI = 206.835 - (alpha*84.6) - (beta*1.015);
        FKGLI = (alpha*11.8) + (beta*0.39) - 15.9;
        FRI = (double)Math.round(FRI);
        FKGLI = (double)(Math.round(FKGLI*10)/10);

        System.out.println("Word #: " + wordNum + "     Syllable #: " + syllNum        //Displays daya
                            + "     Sentences #: " + sentNum);
        System.out.println("Flesch Readability Index: " + FRI);
        System.out.println("Flesch-Kincaid Grade Level Index: " + FKGLI);
    }

    static void checkFile(int args)             //Checks if filename was entered, terminates program if none given
    {
        if(args == 0)
        {
            System.out.println("Error - No File Entered.");
            System.exit(0);
        }
    }

    static boolean isWord(String word)          //Checks if a string contains alphabetical lettes
    {
        for(int i = 0; i < word.length(); i++)
        {
            if(Character.isLetter(word.charAt(i)))
                return true;
        }
        return false;
    }

    static void countWords(int wordData[], String fileName)    //Counts all the data
    {
        try
        {
            File infile = new File(fileName);       //Opens file
            Scanner sc = new Scanner(infile);
            String line;
            String wordArray[];
            while(sc.hasNextLine())             //Increments line by line though the file
            {
                line = sc.nextLine();
                wordArray = line.split(" ");    //Turns line into an array of words
                for(int i = 0; i < wordArray.length; i++)       //Increments though the word array
                {
                    if(isWord(wordArray[i]))            //Counts syllables if string counts as word
                    {
                        wordData[0]++;
                        wordData[1] += countSyllables(wordArray[i]);
                    }
                    if(isSentenceEnd(wordArray[i]))     //Checks if there is sentece ending punctuation attached to word
                    {
                        wordData[2]++;
                    }
                }
            }
        }
        catch(FileNotFoundException ex)
        {
            System.out.println("Error - Invalid File Name Entered");
            System.exit(0);
        }
    }

    static int countSyllables(String word)
    {
        int syllCount = 0;
        boolean volAdj = false;
        char lastvowel = 0;
        for(int i = 0; i < word.length(); i++)       //Increments though ch in the word
        {
            if(isVowel(word.charAt(i)))
                lastvowel = word.charAt(i);
            if(volAdj == false)                 //Checks if encountered a vowel, if not then check for vowel
            {
                if(isVowel(word.charAt(i)))
                    volAdj = true;
            }
            else                                //If encountered vowel
            {
                if(!isVowel(word.charAt(i)) && isWord(Character.toString(word.charAt(i))))        //If adjacent vowels end, then increment syllCount
                {
                    syllCount++;
                    volAdj = false;
                }
            }
        }
        lastvowel = Character.toLowerCase(lastvowel);
        if(volAdj == true && lastvowel != 'e')   //Makes sure words that end in 'e' don't get an extra syllable
            syllCount++;
        if(syllCount == 0)              //Makes sure all words have at least one syllable
            syllCount++;
        return syllCount;
    }

    static boolean isVowel(char ch)             //Checks if ch is a vowel
    {
        char c = Character.toLowerCase(ch);
        if(c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u' || c == 'y')
            return true;
        return false;
    }

    static boolean isSentenceEnd(String word)       //Checks if a string has ending punctuation
    {
        char c;
        for(int i = 0; i < word.length(); i++)
        {
            c = word.charAt(i);
            if(c == '.' || c == ':' || c == ';' || c == '?' || c == '!')
                return true;
        }
        return false;
    }
}

File: textanalyzer.cpp
//Mark Gomes
//CSC 330
//Assignment 1 - Text Parsing
//9/25/17
#include <iostream>
#include <sstream>
#include <string>
#include <fstream>
#include <cstdlib>
#include <stdio.h>
#include <ctype.h>
#include <iomanip>
#include <math.h>

using namespace std;

void checkFile(int argc);           //Checks if a filename was given
bool isWord(string word);           //Checks if a string contains an alphabetical character
void countWords(int worddata[], string filename);       //Counts words in a string
int countSyllables(string word);            //Counts syllables in a string
bool isVowel(char ch);                      //Checks if a character is a vowel
bool isSentenceEnd(string word);            //Checks if a string contains ending punctution

int main(int argc, char* argv[])
{
    checkFile(argc);                        //Checks if any filename is given
    string filename = argv[1];
    float wordNum, syllNum, sentNum;
    float alpha, beta, FRI, FKGLI;
    int worddata[3];                        //0: Words; 1: Syllables; 2: Sentences
    worddata[0] = 0;    worddata[1] = 0;    worddata[2] = 0;
    countWords(worddata, filename);
    wordNum = worddata[0];  syllNum = worddata[1];  sentNum = worddata[2];

    alpha = syllNum/wordNum;                //Calculate indexes and display information
    beta = wordNum/sentNum;
    FRI = 206.835 - (alpha*84.6) - (beta*1.015);
    FKGLI = (alpha*11.8) + (beta*0.39) - 15.9;
    cout << "Word #: " << wordNum << "    " << "Syllables #: " << syllNum << "    " << "Sentence #: " << sentNum << endl;
    cout << "Flesch Readability Index: " << round(FRI) << endl;
    FKGLI = round(FKGLI * 10) / 10;                //Rounds to one decimal points
    cout << "Flesch-Kincaid Grade Level Index: " << setprecision(2) << FKGLI << endl;
    return 0;
}

void checkFile(int argc)
{
    if(argc == 1)
    {
        cout << "Error - No File Entered." << endl;
        exit(1);
    }
}

bool isWord(string word)
{
    bool hasLetter = false;
    for(int i = 0; i < word.length(); i++)
    {
        if(isalpha(word[i]))
                hasLetter = true;
    }
    return hasLetter;
}

void countWords(int worddata[], string filename)
{
    ifstream infile;
    infile.open(filename.c_str());
    if(!infile.is_open())                           //Checks if file is available
    {
        cout << "Error - Invalid File Name Entered" << endl;
        exit(1);
    }

    int wordNum = 0;
    string word;
    int endlpos;
    while(infile.peek() != EOF)                     //Iterates through file until EOF(end of file)
    {
        getline(infile, word, ' ');                 //Stores strings from file between every space
        endlpos = word.find("\n");
        if(endlpos != -1)                           //Checks if it's two words becuase of nextline
        {
            if(isWord(word.substr(endlpos)))
            {
                wordNum++;
            }
        }
        if(isWord(word))                            //Counts syllables if string counts as word
        {
            wordNum++;
            worddata[1] += countSyllables(word);
        }
        if(isSentenceEnd(word))                     //Check if there is sentencing ending punctuation attached to word
            worddata[2]++;
    }
    worddata[0] = wordNum;
}

int countSyllables(string word)
{
    int syllCount = 0;
    bool volAdj = false;
    char lastvowel;
    stringstream ss;
    string letter;
    for(int i = 0; i < word.length(); i++)
    {
        if(isVowel(word[i]))            //Records the last vowel
            lastvowel = word[i];   
 
        if(volAdj == false)             //If not yet encountered vowel then check for one
        {
            if(isVowel(word[i]))
                volAdj = true;
        }
        else                            //If already encountered vowels
        {
            ss << word[i];
            ss >> letter;
            if(!isVowel(word[i]) && isWord(letter))       //If adjacent vowels end then increment syllnum
            {
                syllCount++;
                volAdj = false;
            }
        }
    }
    lastvowel = tolower(lastvowel);
    if(volAdj == true && lastvowel != 'e')          //Prevents words with ending 'e' to have extra syllable
        syllCount++;
    if(syllCount == 0)              //Makes sure all words have at least 1 syllable
        syllCount++;
    return syllCount;
}

bool isVowel(char ch)
{   
    bool vowel = false;
    char c;
    c = tolower(ch);
    if(c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u' || c == 'y')
        vowel = true;
    return vowel;
}
bool isSentenceEnd(string word)
{
    bool sentEnd = false;
    char c;
    for(int i = 0; i < word.length(); i++)
    {
        c = word[i];
        if(c == '.' || c == ':' || c == ';' || c == '?' || c == '!')
            sentEnd = true;
    }
    return sentEnd;
}
